src/pdaaal/engine/UntypedPAutomaton.cpp

```diff
@@ -0,0 +1,541 @@
+/*
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ *  Copyright Morten K. Schou
+ */
+
+/*
+ * File:   UntypedPAutomaton.cpp
+ * Author: Morten K. Schou <morten@h-schou.dk>
+ *
+ * Created on 21-01-2020.
+ */
+
+
+#include "UntypedPAutomaton.h"
+#include <stack>
+#include <boost/functional/hash.hpp>
+
+namespace pdaaal {
+
+    /*
+     * PreStar, PostStar *
+     */
+    void UntypedPAutomaton::_pre_star() {
+        // This is an implementation of Algorithm 1 (figure 3.3) in:
+        // Schwoon, Stefan. Model-checking pushdown systems. 2002. PhD Thesis. Technische Universität München.
+        // http://www.lsv.fr/Publis/PAPERS/PDF/schwoon-phd02.pdf (page 42)
+
+        std::vector<std::unique_ptr<temp_edge_t>> edges;
```

Comment:
Use direct indexing instead of pointers - it avoids double-heap lookups and you get more cache-friendly:
```
std::vector<temp_edge_t> edges;
std::stack<size_t> trans;
std::vector<temp_edge_t> rel;
```